gateway/downlink.go

```diff
@@ -0,0 +1,210 @@
+package gateway
+
+/*
+#cgo CFLAGS: -I./sx1302/libloragw/inc -I./sx1302/libtools/inc
+#cgo LDFLAGS: -L./sx1302/libloragw -lloragw -L./sx1302/libtools -lbase64 -lparson -ltinymt32  -lm
+
+#include "../sx1302/libloragw/inc/loragw_hal.h"
+#include "gateway.h"
+#include <stdlib.h>
+
+*/
+import "C"
+
+import (
+	"context"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/viam-modules/gateway/node"
+	"go.thethings.network/lorawan-stack/v3/pkg/crypto"
+	"go.thethings.network/lorawan-stack/v3/pkg/types"
+	"go.viam.com/utils"
+)
+
+const (
+	joinDelay     = 6         // rx2 delay in seconds for sending join accept message.
+	downlinkDelay = 2         // rx2 delay in seconds for downlink messages.
+	rx2Frequency  = 923300000 // Frequency to send downlinks on rx2 window, lorawan rx2 default
+	rx2SF         = 12        // spreading factor for rx2 window, default for lorawan
+	rx2Bandwidth  = 0x06      // 500k bandwidth, default bandwidth for downlinks
+)
+
+func (g *gateway) sendDownlink(ctx context.Context, payload []byte, isJoinAccept bool, packetTime time.Time) error {
+	txPkt := C.struct_lgw_pkt_tx_s{
+		freq_hz:     C.uint32_t(rx2Frequency),
+		freq_offset: C.int8_t(0),
+		// tx_mode 0 is immediate, 1 for timestampted with count_us delay
+		// doing immediate mode with sleep to exit on context cancelation.
+		tx_mode:    C.uint8_t(0),
+		rf_chain:   C.uint8_t(0),
+		rf_power:   C.int8_t(26),            // tx power in dbm
+		modulation: C.uint8_t(0x10),         // LORA modulation
+		bandwidth:  C.uint8_t(rx2Bandwidth), // 500k
+		datarate:   C.uint32_t(rx2SF),
+		coderate:   C.uint8_t(0x01), // code rate 4/5
+		invert_pol: C.bool(true),    // Downlinks are always reverse polarity.
+		size:       C.uint16_t(len(payload)),
+		preamble:   C.uint16_t(8),
+		no_crc:     C.bool(true), // CRCs in uplinks only
+		no_header:  C.bool(false),
+	}
+
+	var cPayload [256]C.uchar
+	for i, b := range payload {
+		cPayload[i] = C.uchar(b)
+	}
+	txPkt.payload = cPayload
+
+	var waitDuration time.Duration
+	switch isJoinAccept {
+	case true:
+		// 47709/32*time.Microsecond is the internal delay of sending a packet
+		waitDuration = (joinDelay * time.Second) - (time.Since(packetTime)) - 47709/32*time.Microsecond
+	default:
+		waitDuration = (downlinkDelay * time.Second) - (time.Since(packetTime)) - 47709/32*time.Microsecond
+	}
```

Comment:
If else with a bool is more idiomatic.

```suggestion
        // 47709/32*time.Microsecond is the internal delay of sending a packet
        waitDuration := (downlinkDelay * time.Second) - (time.Since(packetTime)) - 47709/32*time.Microsecond
	if isJoinAccept {
		waitDuration = (joinDelay * time.Second) - (time.Since(packetTime)) - 47709/32*time.Microsecond
	}
```